gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java

```diff
@@ -2032,9 +2034,47 @@ public GoogleCloudStorageItemInfo composeObjects(
   <RequestT extends StorageRequest<?>> RequestT configureRequest(
       RequestT request, String bucketName) {
     setRequesterPaysProject(request, bucketName);
+
+    if (request instanceof Storage.Objects.Get || request instanceof Storage.Objects.Insert) {
+      setEncryptionHeaders(request);
+    }
+
+    if (request instanceof Storage.Objects.Rewrite || request instanceof Storage.Objects.Copy) {
+      setEncryptionHeaders(request);
+      setDecryptionHeaders(request);
+    }
+
     return request;
   }
 
+  private <RequestT extends StorageRequest<?>> void setEncryptionHeaders(RequestT request) {
+    HttpHeaders headers = request.getRequestHeaders();
+    if (storageOptions.getEncryptionAlgorithm() != null && storageOptions.getEncryptionKey() != null
```

Comment:
Add this validation to options class - if customer will specify only one property we should fail Options class creation.